### PR TITLE
feat/v8

### PR DESCRIPTION
**Pull request by dannyhw**
- updated to be compatible with v8
- new ui
- ui components extracted to ui package
- better hmr (story list updates live without reloads)

breaking:
- new requirements
  - @gorhom/bottom-sheet, react-native-gesture-handler, react-native-reanimated, react-native-svg 

todo:

- [x] new controls ui
- [x] currently on rn cli a react-native-url-polyfill is needed to not crash the app, need to track this down and fix
- [x] rework the bottom sheet
- [x] make the new ui work on web/wide screens
- [x] refactor theme, right now its a mix between old and new (copy of web theme). Instead should update anywhere that references old theme values to use new theme values.

progress:

https://github.com/storybookjs/react-native/assets/3481514/be142218-189f-47f4-a3b6-13856a78c836

### CODE REVIEW

File: packages/ondevice-controls/src/ControlsPanel.tsx [hunk 1]
```diff
@@ -74,10 +76,23 @@ const ControlsPanel = ({ api }: { api: API }) => {
 
   const [argsFromHook, updateArgs, resetArgs] = useArgs(storyId, store);
 
+  const [story, setStory] = useState<Record<string, any> | undefined>(undefined);
+
+  useEffect(() => {
+    const getStory = async () => {
+      const thisStory = await store.fromId(storyId);
```
Comment (dannyhw) [inline packages/ondevice-controls/src/ControlsPanel.tsx]: async is making this awkward

File: packages/react-native/src/View.tsx [hunk 2]
```diff
@@ -169,14 +175,17 @@ export class View {
...
       store: () => ({
-        fromId: (id) =>
-          this._preview.storyStore.getStoryContext(this._preview.storyStore.fromId(id)),
+        fromId: async (id) => {
+          // return this._preview.storyStore.getStoryContext(this._preview.storyStore.fromId(id));
+
+          // console.log()
+          // this._preview.store
+          // console.log('fromId', id);
+          return this._preview.loadStory({ storyId: id });
```
Comment (dannyhw) [inline packages/react-native/src/View.tsx]: @shilman @tmeasday this loadStory is async and makes it hard to work with. Also its just not returning the story for some reason. Are we missing something?

**Comment by dannyhw** [comment]
@SocketSecurity ignore-all

File: packages/react-native/src/View.tsx (hunk 2 above)
Comment (shilman) [reply packages/react-native/src/View.tsx]:
```
return this._preview.getStoryContext(this._idToPrepared[id]);
```

**Comment by vstlouis** [comment]
Hey! Great work here. Just wondering what work is left to get v8 support and maybe offer assistance to move this along. Thanks!

**Comment by dannyhw** [comment]
@vstlouis Hey at this point its missing

- [x] Updated controls ui 
- [x] rework the bottom sheet
- [ ] make the new ui work on web/wide screens
- [x] refactor theme, right now its a mix between old and new (copy of web theme). Instead should update anywhere that references old theme values to use new theme values.


Problem is I've had a lot going on in my personal/work life so haven't had any time to pick this up.

If you or anyone wants to have a go at bringing the controls ui closer to the web version of mobile ui or anything else listed then im happy to review prs

**Comment by dannyhw** [comment]
v8.2 was a bit of a setback as there was a large change to upstream apis that broke a lot in RN storybook. I've now introduced
`@storybook/react-native/metro/withStorybook`
That works around many issues and also provides the required metro config.

Metro config for storybook should now be done like this in 8.2.0-alpha.1

```
const withStorybook = require('@storybook/react-native/metro/withStorybook');

module.exports = withStorybook(defaultConfig, {
  enabled: process.env.STORYBOOK_ENABLED === 'true',
  configPath: path.resolve(__dirname, './.storybook'),
});

```

This follows the same pattern as tamagui and nativewind for metro configuration. Should also be compatible, see this tamagui example: https://github.com/dannyhw/tamagui-expo-router-sb8

Note that this is still early and until we reach RC status there can still be a few more changes. I believe this withStorybook change is actually positive and makes it easier to explain setup anyway.

**Comment by kzlar** [comment]
> v8.2 was a bit of a setback as there was a large change to upstream apis that broke a lot in RN storybook. I've now introduced `@storybook/react-native/metro/withStorybook` That works around many issues and also provides the required metro config.
> 
> Metro config for storybook should now be done like this in 8.2.0-alpha.1
> 
> ```
> const withStorybook = require('@storybook/react-native/metro/withStorybook');
> 
> module.exports = withStorybook(defaultConfig, {
>   enabled: process.env.STORYBOOK_ENABLED === 'true',
>   configPath: path.resolve(__dirname, './.storybook'),
> });
> ```
> 
> This follows the same pattern as tamagui and nativewind for metro configuration. Should also be compatible, see this tamagui example: https://github.com/dannyhw/tamagui-expo-router-sb8
> 
> Note that this is still early and until we reach RC status there can still be a few more changes. I believe this withStorybook change is actually positive and makes it easier to explain setup anyway.

Hey @dannyhw , wondering how this ties in to existing metro configs that select different App.tsx files based on STORYBOOK_ENABLED eg:
```
if (process.env.STORYBOOK_ENABLED === 'true') {
  const { generate } = require('@storybook/react-native/scripts/generate')
  generate({
    // eslint-disable-next-line no-undef
    configPath: path.resolve(__dirname, './.storybook'),
  })
  config.resolver.sourceExts = [
    // XXX: For the App.storybook.tsx file this must be first
    'storybook.tsx',
    ...config.resolver.sourceExts,
    'mjs',
  ]
}
```
The idea is to make sure storybook isn't bundled on prod apps.
p.s. appreciate the effort on updating this to 8.2 :)

**Comment by dannyhw** [comment]
@kzlar  If you want to remove storybook from the bundle the easiest way is to adjust metros resolveRequest to ignore storybook files.

```
defaultConfig.resolver.resolveRequest = (context, moduleName, platform) => {
  const defaultResolveResult = context.resolveRequest(
    context,
    moduleName,
    platform
  );

  if (
    process.env.STORYBOOK_ENABLED !== "true" &&
    defaultResolveResult?.filePath?.includes?.(".storybook/")
  ) {
    return {
      type: "empty",
    };
  }

  return defaultResolveResult;
};
```

withStorybook should respect any existing resolveRequest function.

I could add this automatically but it requires resetting cache when switching between including/not including storybook, so I'm not sure.

**Comment by kzlar** [comment]
> @kzlar If you want to remove storybook from the bundle the easiest way is to adjust metros resolveRequest to ignore storybook files.
> 
> ```
> defaultConfig.resolver.resolveRequest = (context, moduleName, platform) => {
>   const defaultResolveResult = context.resolveRequest(
>     context,
>     moduleName,
>     platform
>   );
> 
>   if (
>     process.env.STORYBOOK_ENABLED !== "true" &&
>     defaultResolveResult?.filePath?.includes?.(".ondevice/")
>   ) {
>     return {
>       type: "empty",
>     };
>   }
> 
>   return defaultResolveResult;
> };
> ```
> 
> withStorybook should respect any existing resolveRequest function.
> 
> I could add this automatically but it requires resetting cache when switching between including/not including storybook, so I'm not sure.

@dannyhw ty for your reply, and please let me know if you want further discussion on discord or anywhere else so to not spam this pr :)

Wouldn't that method above still include storybook in the production app? The App.storybook.tsx file has this import for me:
```
import { view } from './.storybook/storybook.requires'

const StorybookApp = view.getStorybookUI({
  storage: {
    getItem: AsyncStorage.getItem,
    setItem: AsyncStorage.setItem,
  },
})
```
If I merge it into App.tsx and add the resolveRequest as you mentioned, it'd still bundle in everything imported from `storybook.requires.ts` which has a bunch of stuff that aren't useful in production. Right?

Another minor issue with the above is that my stories are shared between web and native in a package that is separate from my expo app (still using a relatively old tamagui monorepo setup with apps/next, apps/expo for the apps and packages/app for the components).

**Comment by dannyhw** [comment]
@kzlar it wouldn't bundle it in because its resolving an empty module, you'd basically have a fake empty module in its place. You can also add other filepaths here if you want and exclude everything that you think shouldn't get included.

see
```
  if (
    process.env.STORYBOOK_ENABLED !== "true" &&
    defaultResolveResult?.filePath?.includes?.(".storybook/")
  ) {
    return {
      type: "empty",
    };
  }
  ```
  
  Feel free to dm me on discord

**Comment by dannyhw** [comment]
added an alpha version of the expo template

to try run `npx create-expo-app --template expo-template-storybook@next alphastorybook8`

**Comment by moimael** [comment]
Hi! I'm getting 

```iOS Bundling failed 11974ms
The package at "../../node_modules/jose/dist/node/esm/runtime/decrypt.js" attempted to import the Node standard library module "node:crypto".
It failed because the native React runtime does not include the Node standard library.
```

on `8.2.0-alpha.1`

It randomly alternates between `node:crypto` and `node:buffer` not found in `jose`. I tried digging but have no idea which package pulls it and why.

When storybook is disabled the app builds just fine. And storybook works fine in v7. 

Any pointers?

Thanks!

**Comment by dannyhw** [comment]
Sorry that this is dragging out still, I'm going to start putting more time into this again

I've started working on fixing the mess with themes and updating controls ui

last thing after that will be the web/widescreen ui and solving the polyfill requirement for rn cli.

If you want to not have issues with polyfills on url/crypto I recommend using expo though since the important things are polyfilled by default.

**Comment by dannyhw** [comment]
dark mode 😄

image

image

**Comment by dannyhw** [comment]
New alpha version `8.3.0-alpha.0`out with simplfied theme and updated controls ui to (kind of) match web style
`npx create-expo-app --template expo-template-storybook@next sbv8alpha` 

If you have a custom theme then sorry it probably breaks but with the new ui I wanted to reuse the web theme to make it easier to update in the future.

**Comment by dannyhw** [comment]
8.3.0-alpha.1 fixes rncli compatibility by applying the polyfill automatically if URLSearchParams.get is not available

Have it working in my example project here:

https://github.com/dannyhw/react-native-storybook-starter/pull/7

**Comment by dannyhw** [comment]
Ok added a basic widescreen ui and made sure each addon is working on web. Don't expect anything fancy but it at least works.

(you can use react storybook with addon-react-native-web to get the full storybook web experience)

image

**Comment by dannyhw** [comment]
8.3.0-alpha.2 is out with more fixes and all the other changes

expo template on next tag also updated
`npx create-expo-app --template expo-template-storybook@next sbv8alpha`

**Pull request merged**